### PR TITLE
Enable controller to be aware of number of available IP addresses

### PR DESCRIPTION
**Pull request by liwenwu-amazon**
This PR addresses issue #18 .

It defines **eks.amazon.com/ip** as a [kubernetes' extended resources](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#extended-resources) to represent the number of available of available IP addresses on each node.

### CODE REVIEW

**Comment by mattlandis** [comment]
This plugin is usable outside of an eks cluster so the resource should not be scoped to eks.  Maybe vpc-cni.amazon.com/ip or vpc-cni.aws.amazon.com/ip?

**Comment by aaithal** [comment]
I second @mattlandis  comment here. Also, if you have 3611 files in a single commit, it's impossible for anyone to review it effectively.

**Comment by ofiliz** [comment]
I am not sure if this resource would ever make sense outside of EKS. If we name it generically, we risk confusing people who do not need it into thinking they do. This resource exists only because a specific implementation detail of vpc-cni-k8s plugin requires it. Other CNI plugins (at least as of today) do not need it.

Also, the name of the resource should be something like "private-ipv4-address". "ip" is the name of the protocol, and even if it is commonly (and incorrectly) understood to be an address, let's be clear on what type of address it really is.

**Comment by bmoylan** [comment]
We deploy a custom-configured Kubernetes cluster on EC2 and would love to have this feature generically available instead of tied directly to EKS.

**Comment by liwenwu-amazon** [comment]
how about `vpc.amazonaws.com/ip`?

**Comment by aaithal** [comment]
how about `vpc.amazonaws.com/ipv4` to state explicitly that this is for ipv4 addresses?

File: Gopkg.toml [hunk 1]
```diff
@@ -21,13 +21,17 @@
 #  version = "2.4.0"
 
 
+[[constraint]]
+  name = "github.com/aws/amazon-ecs-agent"
```
Comment (samuelkarp) [inline Gopkg.toml]: Why did you vendor the ECS agent?

File: Gopkg.lock [hunk 2]
```diff
@@ -16,8 +16,8 @@
 [[projects]]
   name = "github.com/aws/amazon-ecs-agent"
   packages = ["agent/utils/ttime","agent/utils/ttime/mocks"]
-  revision = "ea58fddf8462034ff2034aa9788b02341d7702e3"
-  version = "v1.15.2"
+  revision = "ea383f2ff70290029ee69c7089160b10941baeb6"
+  version = "v1.16.1"
 
 [[projects]]
   branch = "master"
```
Comment (samuelkarp) [inline Gopkg.lock]: Why did you vendor github.com/aws/amazon-ecs-cni-plugins?

File: Gopkg.toml (hunk 1 above)
Comment (liwenwu-amazon) [reply Gopkg.toml]: Right now, it uses some code from amazon-ecs-agent and amazon-ecs-cni-plugins. We are planning to move these shared code out into another library repo and only vendor this new library repo.

Comment (samuelkarp) [reply Gopkg.toml]: What code are you using?  It's worth evaluating whether to factor it out or to just stop using the code; not everything in the ECS agent is code that we'd recommend reusing.

Comment (liwenwu-amazon) [reply Gopkg.toml]: opened issue #27 to track this.   Thanks

**Comment by liwenwu-amazon** [comment]
Move this to a new PR #31  where it is broken down into multiple commit

**Pull request closed without merging**